runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1536 @@
+		J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmindexOffset, (U_64)vmindex);
+		J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmtargetOffset, (U_64)target);
+		Trc_JCL_java_lang_invoke_MethodHandleNatives_initImpl_setData(currentThread, flags, clazzObject, vmindex, target);
+	}
+}
+
+static char *
+sigForPrimitiveOrVoid(J9JavaVM *vm, J9Class *clazz)
+{
+	char result = 0;
+	if (clazz == vm->booleanReflectClass) {
+		result = 'Z';
+	} else if (clazz == vm->byteReflectClass) {
+		result = 'B';
+	} else if (clazz == vm->charReflectClass) {
+		result = 'C';
+	} else if (clazz == vm->shortReflectClass) {
+		result = 'S';
+	} else if (clazz == vm->intReflectClass) {
+		result = 'I';
+	} else if (clazz == vm->longReflectClass) {
+		result = 'J';
+	} else if (clazz == vm->floatReflectClass) {
+		result = 'F';
+	} else if (clazz == vm->doubleReflectClass) {
+		result = 'D';
+	} else if (clazz == vm->voidReflectClass) {
+		result = 'V';
+	}
+
+	if (result) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		char* signature = (char*)j9mem_allocate_memory(2, OMRMEM_CATEGORY_VM);
+		signature[0] = result;
+		signature[1] = '\0';
+		return signature;
+	}
+	return NULL;
+}
+
+char *
+getClassSignature(J9VMThread *currentThread, J9Class * clazz)
+{
+	PORT_ACCESS_FROM_JAVAVM(currentThread->javaVM);
+
+	U_32 numDims = 0;
+
+	J9Class * myClass = clazz;
+	while (J9ROMCLASS_IS_ARRAY(myClass->romClass)) {
+		J9Class * componentClass = (J9Class *)(((J9ArrayClass*)myClass)->componentType);
+		if (J9ROMCLASS_IS_PRIMITIVE_TYPE(componentClass->romClass)) {
+			break;
+		}
+		numDims++;
+		myClass = componentClass;
+	}
+
+	J9UTF8 * romName = J9ROMCLASS_CLASSNAME(myClass->romClass);
+	U_32 len = J9UTF8_LENGTH(romName);
+	char * name =  (char *)J9UTF8_DATA(romName);
+	U_32 length = len + numDims;
+	if (* name != '[') {
+		length += 2;
+	}
+
+	length++; //for null-termination
+	char * sig = (char *)j9mem_allocate_memory(length, OMRMEM_CATEGORY_VM);
+	U_32 i;
+	for (i = 0; i < numDims; i++) {
+		sig[i] = '[';
+	}
+
+	if (*name != '[') {
+		sig[i++] = 'L';
+	}
+
+	memcpy(sig+i, name, len);
+	i += len;
+
+	if (*name != '[') {
+		sig[i++] = ';';
+	}
+	sig[length-1] = '\0';
+
+	return sig;
+}
+
+char *
+getSignatureFromMethodType(J9VMThread *currentThread, j9object_t typeObject, UDATA *signatureLength)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	j9object_t ptypes = J9VMJAVALANGINVOKEMETHODTYPE_PTYPES(currentThread, typeObject);
+	U_32 numArgs = J9INDEXABLEOBJECT_SIZE(currentThread, ptypes);
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	char** signatures = (char**)j9mem_allocate_memory((numArgs + 1) * sizeof(char*), OMRMEM_CATEGORY_VM);
+	*signatureLength = 2; //space for '(', ')'
+	for (U_32 i = 0; i < numArgs; i++) {
+		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+		signatures[i] = sigForPrimitiveOrVoid(vm, pclass);
+		if (!signatures[i]) {
```

Comment:
Needs to handle the NativeOOM case as well - likely by checking for a pending exception